src/compiler/checker.ts

```diff
@@ -9042,6 +9050,33 @@ namespace ts {
                     const nonNeverTypes = filter((<IntersectionType>objectType).types, t => !isMappedTypeToNever(t));
                     return type.simplified = getSimplifiedType(getIndexedAccessType(getIntersectionType(nonNeverTypes), type.indexType));
                 }
+
+                // Given an indexed access type T[K], if T is an intersection containing a generic type and one or
+                // more mapped types which map over the keys of that generic, '(T & {[P in keyof T]: Foo})[K]', return a
+                // transformed type that moves the top-level intersection into the mapped type: '{[P in keyof T]: T[P] & Foo}[K]'
+                // If we don't do such a transformation now, due to how we reason over intersections, we will never come to the
+                // realization that the member we're trying to pluck out is influenced by both `T[P]` _and_ the mapped type template
+                const isInSet = isMappedTypeOverKeyofGenericInSet((<IntersectionType>objectType).types);
+                if (some((<IntersectionType>objectType).types, isInSet)) {
+                    for (const t of (<IntersectionType>objectType).types) {
+                        if (isInSet(t)) {
+                            const constraint = getConstraintTypeFromMappedType(t as MappedType);
+                            let matchIndex: number | undefined;
+                            const match = forEach((<IntersectionType>objectType).types, (t, i) => getIndexType(t) === constraint ? (matchIndex = i, t) : undefined)!;
+                            const indexedMatch = getIndexedAccessType(match, getTypeParameterFromMappedType(t as MappedType));
+                            const newTemaplate = getIntersectionType([indexedMatch, getTemplateTypeFromMappedType(t as MappedType)]);
+                            const newMappedType = <MappedType>createObjectType(ObjectFlags.Mapped, t.symbol);
+                            newMappedType.declaration = (t as MappedType).declaration;
+                            newMappedType.templateType = newTemaplate;
+                            newMappedType.typeParameter = (t as MappedType).typeParameter;
+                            newMappedType.constraintType = (t as MappedType).constraintType;
+                            const newSet = (<IntersectionType>objectType).types.slice();
+                            newSet.splice(matchIndex!, 1); // Remove the generic
+                            newSet.splice(newSet.indexOf(t), 1, newMappedType); // Remove the original mapped type

```

Comment:
Why not just index at the original location?

Reply:
... fair point. I was in a splicy mood, I suppose.